core/tracestack.lua

```diff
@@ -0,0 +1,214 @@
+  -- Must be popped with `pop(returnOfPush)`.
+  pushContent = function(self, content, tag, line, column, options, file)
+    if type(content) ~= "table" then
+      SU.warn("Content parameter of SILE.traceStack:pushContent must be a table", true)
+      content = {}
+    end
+    tag = tag or content.tag
+    if not tag then
+      SU.warn("Tag should be specified or inferable for SILE.traceStack:pushContent", true)
+    end
+    return self:_pushFrame({
+        tag = tag or "???",
+        file = file or content.file or SILE.currentlyProcessingFile,
+        line = line or content.line,
+        column = column or content.col,
+        options = options or content.attr or {}
+      })
+  end,
+
+  -- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+  -- Must be popped with `pop(returnOfPush)`.
+  pushText = function(self, text, line, column, file)
+    return self:_pushFrame({
+        text = text,
+        file = file,
+        line = line,
+        column = column
+      })
+  end,
+
+  -- Internal: Push complete frame
+  _pushFrame = function (self, frame)
+    frame.toLocationString = self._frameToLocationString
+    frame.typesetter = SILE.typesetter
+    -- Push the frame
+    if SU.debugging("commandStack") then
+      print(string.rep(" ", #self) .. "PUSH(" .. frame:toLocationString(false, true) .. ")")
+    end
+    self[#self + 1] = frame
+    self.afterFrame = nil
+    self._lastPushId = self._lastPushId + 1
+    frame.pushId = self._lastPushId
+    return self._lastPushId
+  end,
+
+  -- Pop previously pushed command from the stack.
+  -- Return value of `push` function must be provided as argument to check for balanced usage.
+  pop = function(self, pushId)
+    if type(pushId) ~= "number" then
+      SU.error("SILE.traceStack:pop's argument must be the result value of the corresponding push", true)
+    end
+    -- First verify that push/pop is balanced
+    local popped = self[#self]
+    if popped.pushId ~= pushId then
+      local message = "Unbalanced content push/pop"
+      local debug = SILE.traceback or SU.debugging("commandStack")
+      if debug then
+        message = message .. ". Expected " .. popped.pushId .. " - (" .. popped:toLocationString() .. "), got " .. pushId
+      end
+      SU.warn(message, debug)
+    else
+      -- Correctly balanced: pop the frame
+      self.afterFrame = popped
+      self[#self] = nil
+      if SU.debugging("commandStack") then
+        print(string.rep(" ", #self) .. "POP(" .. popped:toLocationString(false, true) .. ")")
+      end
+    end
+  end,
+
+  -- Returns frame stack represented by SILE.traceStack and self.afterFrame, but without frames whose typesetter
+  -- is different than what is given as a first parameter (nil = use current typesetter for filtering).
+  -- Most relevant object is in the last index.
+  filteredByTypesetter = function(self, typesetter)
```

Comment:
As far as I can see, this function is no longer called at all anywhere. Does it still have any utility? If not I'll remove it, I'm already refactoring other functions a little so don't worry about doing it @Darkyenus, I'm just looking for feedback to make sure I'm not missing something.

Reply:
It is not used. I have kept it to ease possible future merger with breadcrumb system, as I understand that separate traces per typesetter are an important feature of it.

If you decide that it is no longer necessary, feel free to remove it. You can then also remove [`frame.typesetter = SILE.typesetter` in `_pushFrame`](https://github.com/simoncozens/sile/pull/603/commits/edafe1e9ea16ba743b1fde139a06d38546874ee0#diff-8e3ebbe41b16928909ff53e84982f1d1R331) as that exists only to support this function.

Reply:
Roger that. Working on that now.